### PR TITLE
Merge Coreclr support into FSharp Mainline

### PR DESCRIPTION
**Pull request by KevinRansom**
Okay folks,

this is a monster PR.  It adds support for a CORECLR hosted  compilation, it has preliminary FSI support.  The compiler is in pretty great shape and is used to bootstrap the OSS build.  The main codepaths and test codepaths are Windows.  It still doesn't enable the repo to be built on Mac and Linux, although that is a goal.

@enricosada,
I had to revert the fs.srgen changes because the current package does not build coreclr compliant code.  Please update the package and resubmit these changes.

The code review is too huge to view on GitHub ... sorry about that.

I'm not sure how best to review this:  make a branch download the PR and use VS Code or devenv or your favourite git diff tool to view the delta.

So here is what I did, I hate having to do it:

1.  checkout master branch, pull to get the latest stuff
2.  then xcopy visualfsharp /s into my tempdirectory
3.  create a branch for the pr
4.  download the pr into that new branch
5.  xcopy it into a separate directory in my temp directory
6.  used beyond compare to do a folder comparison.

Add specific code review feedback to this thread.   Please reserve this thread for specific code review feedback.

Thanks

Kevin

### CODE REVIEW

**Comment by enricosada** [comment]
@KevinRansom fssrgen works on .net core, it's in my other pr, i'll send a pr to coreclr branch asap so one diff less

Complete plaintext diff (you can also do .patch):

https://github.com/Microsoft/visualfsharp/pull/1006.diff

@KevinRansom there are lot of whitespace changes (removed lines, blanks, etc), can we try to minimize diff removing these? it's 1/2 of diff the first files
I am pretty sure we can see the complete diff on github if we do fssrgen and minimize whitespace.

File: src/absil/ilsupp.fs [hunk 1]
```diff
@@ -832,31 +829,33 @@ let unlinkResource (ulLinkedResourceBaseRVA:int32) (pbLinkedResource:byte[]) =
...
                 nResNodes <- nResNodes + 1 ;
-                  
+
     // Ok, all tree leaves are in ResFormatNode structs, and nResNodes ptrs are in pResNodes
     let mutable size = 0
     if nResNodes <> 0 then
         size <- size + ResFormatHeader.Width ; // sizeof(ResFormatHeader)
         for i = 0 to (nResNodes - 1) do
             size <- size + pResNodes.[i].Save(ulLinkedResourceBaseRVA, pbLinkedResource, Unchecked.defaultof<byte[]>, 0) ;
-
```
Comment (enricosada) [inline src/absil/ilsupp.fs]: whitespace changes

File: src/absil/ilsupp.fs [hunk 2]
```diff
@@ -1,46 +1,46 @@
 // Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
-module internal Microsoft.FSharp.Compiler.AbstractIL.Internal.Support 
+module internal Microsoft.FSharp.Compiler.AbstractIL.Internal.Support
 
 
 let DateTime1970Jan01 = new System.DateTime(1970,1,1,0,0,0,System.DateTimeKind.Utc) (* ECMA Spec (Oct2002), Part II, 24.2.2 PE File Header. *)
 let absilWriteGetTimeStamp () = (System.DateTime.UtcNow - DateTime1970Jan01).TotalSeconds |> int
 
-
```
Comment (enricosada) [inline src/absil/ilsupp.fs]: whitespace

File: src/absil/ilwrite.fs [hunk 3]
```diff
@@ -3658,7 +3350,7 @@ let writeILMetadataAndCode (generatePdb,desiredMetadataVersion,ilg,emitTailcalls
                mdtableVersionMinor // minor version of table schemata 
                
                ((if stringsBig then 0x01 else 0x00) |||  // bit vector for heap size 
-                (if guidsBig then 0x02 else 0x00) |||  
+                (if guidsBig then 0x02 else 0x00)   |||
```
Comment (enricosada) [inline src/absil/ilwrite.fs]: whitespace

File: src/absil/ilwrite.fs [hunk 4]
```diff
@@ -4461,74 +4159,76 @@ let writeBinaryAndReportMappings (outfile, ilg, pdbfile: string option, signer:
...
+                if portablePDB then 
+                    WritePortablePdbInfo fixupOverlappingSequencePoints showTimes fpdb pdbData
+                else
+                    WritePdbInfo fixupOverlappingSequencePoints showTimes outfile fpdb pdbData
+#endif
             reportTime showTimes "Generate PDB Info"
-            
-          // Now we have the debug data we can go back and fill in the debug directory in the image 
+
+            // Now we have the debug data we can go back and fill in the debug directory in the image
```
Comment (enricosada) [inline src/absil/ilwrite.fs]: whitespace

File: src/fsharp/ExtensionTyping.fs [hunk 5]
```diff
@@ -143,29 +150,29 @@ module internal ExtensionTyping =
...
-                        match box resolver with 
-                        | null -> ()
-                        | _ -> yield (resolver,ilScopeRefOfRuntimeAssembly)
-                  |   None, _ -> 
-                      () ]
-
-            with :? TypeProviderError as tpe ->
-                tpe.Iter(fun e -> errorR(NumberedError((e.Number,e.ContextualErrorMessage),m)) )                        
-                []
+                try
```
Comment (enricosada) [inline src/fsharp/ExtensionTyping.fs]: all whitespace minus 1 line

File: src/fsharp/FSharp.Build-proto/FSharp.Build-proto.fsproj [hunk 6]
```diff
@@ -16,9 +16,9 @@
     <ProjectGuid>{D8BC791F-C1A9-49DC-9432-0F3090537555}</ProjectGuid>
   </PropertyGroup>
   <ItemGroup>
-	  <FsSrGen Include="..\FSharp.Build\FSBuild.txt">
-		  <Link>FSBuild.txt</Link>
-	  </FsSrGen>
+      <FsSrGen Include="..\FSharp.Build\FSBuild.txt">
+          <Link>FSBuild.txt</Link>
```
Comment (enricosada) [inline src/fsharp/FSharp.Build-proto/FSharp.Build-proto.fsproj]: whitespace

**Comment by dsyme** [comment]
This is great work! The diff isn't as daunting as I feared.

We've got an internal code review going on (because the diff is large we're using an internal tool at least for first iteration) - I've left lots of comments for @KevinRansom there! :)

Don

**Comment by dsyme** [comment]
(I started at the bottom of the diff list and worked up - I've still got to review the files under "absil" in depth though I glanced over them quickly)

File: src/absil/ilsupp.fs (hunk 1 above)
Comment (KevinRansom) [inline src/absil/ilsupp.fs]: As I keep saying to @dsyme ... yeah ... there are whitespace changes.

**Comment by KevinRansom** [comment]
As I said to Don ... I am wondering what I have been working on all this time, given how small the diff is.

I am sorry about the white space deltas but by and large they are about eliminating redundant spaces.  I suppose I should write a tool and have it go over the codebase and rip out all of the redundant stuff.

I suppose even more than that we should do what roslyn does and run a standard formatter over the code.

Kevin

**Comment by isaacabraham** [comment]
@KevinRansom Fantomas?

**Comment by KevinRansom** [comment]
@isaacabraham Yes I was thinking of something like that.  But not anytime soon, got to get this puppy ready first.

**Pull request merged**